src/basic/fileio.c

```diff
@@ -412,10 +414,10 @@ int read_full_virtual_file(const char *filename, char **ret_contents, size_t *re
                         n_retries--;
                 } else {
                         /* Double the buffer size */
-                        if (size > READ_FULL_BYTES_MAX / 2)
+                        if (size > READ_FULL_BYTES_MAX / 2 - 1)
                                 return -E2BIG;
 
-                        size *= 2;
+                        size = size * 2 + 1; /* Stay always one less than page size, so we malloc evenly */
                 }
 
                 buf = malloc(size + 1);
```

Comment:
could optimize this further by doing:

`size = malloc_usable_size(buf)-1;` afterwards.

Reply:
Done.